network-libp2p/src/dispatch/codecs/typed.rs

```diff
@@ -193,6 +193,55 @@ impl MessageCodec {
 
         Ok(())
     }
+
+    fn decode_serialized_message(
+        &mut self,
+        src: &mut BytesMut,
+    ) -> Result<(MessageType, BytesMut), Error> {
+        // Make a cursor, so we later know how many bytes we read
+        let mut c = Cursor::new(&src);
+
+        let header = Header::deserialize(&mut c)?;
+        // Deserializing the header was successful
+        let header_length = c.position() as usize;
+
+        // Drop the cursor, so we can mut-borrow the `src` buffer again.
+        drop(c);
+
+        // Preliminary header check (we can't verify the checksum yet)
+        header.preliminary_check()?;
+
+        // Now parse the body.
+        if src.len() >= header.length as usize {
+            // We have read enough bytes to read the full message
+            let message_type = header.type_id;
+
+            // Get buffer for whole message
+            let frame_size = header.length as usize;
+            let mut data = src.split_to(frame_size);
+
+            // Verify the message (i.e. checksum)
+            self.verify(header.checksum, &mut data).map_err(|e| {
+                log::warn!(
+                    "CRC checksum mismatch for message type {}, error: {}",
+                    message_type,
+                    e
+                );
+                e
+            })?;
+
+            // Skip the header to have only the data
+            data.advance(header_length);
+
+            Ok((MessageType::new(message_type), data))
+        } else {
+            // Header size doesn't match the buffer size
+            log::error!(
+                "Request/response decoding error: Header size doesn't match the buffer size"
```

Comment:
If the condition above isn't rewritten to an equality, maybe it could be:
```suggestion
                "Request/response decoding error: Header indicates larger size than the body has"
```
Also, the comment above is redundant, it just repeats the log message.